### PR TITLE
tests: add env variable to keep test session

### PR DESCRIPTION
**Pull request by marcov**
## Reason for This PR

Add the env variable KEEP_TEST_SESSION to keep the created test session
files tree at the end of a test. This allow re-launching tests on a
previous test session without rebuilding all binaries.
## Description of Changes

conftest.py and integration tests README.

## License Acceptance

By submitting this pull request, I confirm that my contribution is made under
the terms of the Apache 2.0 license.

## PR Checklist

`[Author TODO: Meet these criteria. Where there are two options, keep one.]`  
`[Reviewer TODO: Verify that these criteria are met. Request changes if not]`

- [x] All commits in this PR are signed (`git commit -s`).
- [x] Either this PR is linked to an issue, or, the reason for this PR is
      clearly provided. 
- [x] The description of changes is clear and encompassing.
- [x] Either no docs need to be updated as part of this PR, or, the required
      doc changes are included in this PR. Docs in scope are all `*.md` files
      located either in the repository root, or in the `docs/` directory.
- [x] Either no code has been touched, or, code-level documentation for touched
      code is included in this PR.
- [x] Either no API changes are included in this PR, or, the API changes are
      reflected in `firecracker/swagger.yaml`.
- [x] Either the changes in this PR have no user impact, or, the changes in
      this PR have user impact and have been added to the `CHANGELOG.md` file.
- [x] Either no new `unsafe` code has been added, or, the newly added `unsafe`
      code is unavoidable and properly documented.

### CODE REVIEW

**Comment by marcov** [comment]
@dianpopa: this is the followup of: https://github.com/firecracker-microvm/firecracker/pull/1246#discussion_r343071220

**Comment by aghecenco** [comment]
Hey @marcov, thanks for this PR! At a first glance it looks like the new env var has effect when running the tests directly via `pytest` - could you extend it to work with the container too? The recommended way to run integ tests is `tools/devtool test` - this spins up a Docker container and runs `pytest` inside it. I think the `devtool` script should pick up the env var and propagate it to the container.

**Comment by marcov** [comment]
Thank you @aghecenco, done, and sorry for taking some time.

File: tools/devtool [hunk 1]
```diff
@@ -534,6 +534,7 @@ cmd_test() {
         --security-opt seccomp=unconfined \
         --ulimit core=0 \
         --workdir "$CTR_FC_ROOT_DIR/tests" \
+        -e KEEP_TEST_SESSION="$KEEP_TEST_SESSION" \
```
Comment (acatangiu) [inline tools/devtool]: Just propagating the env variable here is not enough.

The `test_session` dir is configured somewhere in `/tmp`. That dir is perishable and will go away along with the `docker session` (`tools/devtool test` runs the container and ends its session once tests complete).

In the current PR form, this line is not useful.

I suggest you either:
1. Extend `devtool` to be able to actually keep the test session. This implies mounting the `test-session-dir` in the container from somewhere on the host (preferably `/tmp/smth`) the same way the `firecracker` dir is mounted to get the source files.
2. Remove this line to merge this PR to enable support for keeping the test session when running without the container, and open an issue to enhance `devtool`.

Note: One can use the `KEEP_TEST_SESSION` functionality even for testing in the container like so:
```bash
firecracker$ tools/devtool shell -p
[Firecracker devtool] Dropping to a privileged shell prompt ...
[Firecracker devtool] Note: /home/ANT.AMAZON.COM/acatan/code/firecracker is bind-mounted under /firecracker
[Firecracker devtool] You are running as root; any files that get created under /firecracker will be owned by root.
root@4a8b30e393fc:/firecracker# cd tests/
root@4a8b30e393fc:/firecracker/tests# KEEP_TEST_SESSION=1 ; pytest
/// running tests but not deleting the test session at the end

root@4a8b30e393fc:/firecracker/tests# pytest
/// running tests again, only doing incremental rebuild of binaries if any src files changed
```

Comment (marcov) [reply tools/devtool]: Thank you for raising this, @acatangiu. I did not think about that. 
Hopefully it's now better.

**Comment by aghecenco** [comment]
I bumped into an error caused by the default jail root [`/srv/jailer/`](https://github.com/firecracker-microvm/firecracker/blob/9bf9333ad4a3f90a1d759e38eef7057af3462af8/jailer/src/lib.rs#L266) being on a different partition than the default test directory root `/tmp/firecracker_test_session`. The test suite can't hardlink the chroots to the test session dir in this case. You could either:
* check that the test session dir is on the same partition as `/srv/jailer` as the test suite doesn't overwrite it anywhere (risky and unrecommended);
* not persist the jails; persist just the build artefacts.

**Comment by acatangiu** [comment]
@dhrgit is reworking this area right now and will add clean support for keeping the test session as part of #1411 (he said he will post a PR soon).

@marcov I suggest you not spend time to patch the current `devtool` as it will probably be throw-away work. I guess we can keep just the `KEEP_TEST_SESSION` functionality in this PR which will provide enough support in the meantime.

**Comment by marcov** [comment]
Ok, rolled back to the initial version and clarified about the need to be used inside a container. Thanks for the in depth review

> What's with the new TR_TMPDIR env var? 

@aghecenco apparently you can override the default `/tmp/firecracker_test_session`, see conftest.py

**Pull request merged**